plugins/@grouparoo/google-sheets/src/lib/shared/Spreadsheet.ts

```diff
@@ -0,0 +1,252 @@
+  loaded: boolean;
+  docId: string;
+  sheetId: string;
+
+  constructor(credentials: SimpleAppOptions, sheetUrl: string) {
+    this.credentials = credentials;
+    this.connected = false;
+
+    const { docId, sheetId } = parseUrl(sheetUrl);
+    if (!docId) {
+      throw `no spreadsheet document given: ${sheetUrl}`;
+    }
+    this.docId = docId;
+    this.sheetId = sheetId;
+
+    this.doc = new GoogleSpreadsheet(docId);
+    this.sheet = null;
+  }
+
+  async connect() {
+    if (this.connected) {
+      return;
+    }
+    const client_email = this.credentials.client_email;
+    const private_key = this.credentials.private_key
+      ?.toString()
+      .replace(/\\n/g, "\n");
+    await this.doc.useServiceAccountAuth({ client_email, private_key });
+    this.connected = true;
+  }
+
+  async load() {
+    if (this.sheet) {
+      await this.loadHeaders();
+      return this.sheet;
+    }
+    await this.connect();
+    const doc = this.doc;
+    await doc.loadInfo(); // loads document properties and worksheets
+
+    if (this.sheetId) {
+      this.sheet = doc.sheetsById[this.sheetId];
+    }
+    if (!this.sheet) {
+      // didn't have that id, use the first one
+      this.sheet = doc.sheetsByIndex[0];
+    }
+    await this.loadHeaders();
+    return this.sheet;
+  }
+
+  async loadHeaders() {
+    try {
+      await this.sheet.loadHeaderRow();
+    } catch (e) {}
+  }
+
+  async read({ limit, offset }) {
+    const sheet = await this.load();
+    const results = [];
+    if (offset < sheet.rowCount) {
+      const rows = await sheet.getRows({ limit, offset });
+      for (const row of rows) {
+        const result: any = {};
+        result["_rowNumber"] = row.rowNumber;
+        for (const header of sheet.headerValues) {
+          result[header] = row[header];
+        }
+        results.push(result);
+      }
+    }
+    return results;
+  }
+
+  async getHeaders() {
+    await this.load();
+    return this.sheet.headerValues;
+  }
+
+  async _invalidateCache(): Promise<void> {
+    const contentId = `${this.docId}-${this.sheetId}`;
+    await objectCacheInvalidate({ objectId: contentId });
+  }
+
+  async getAllRows() {
+    await this.load();
+    const allRows = [];
+    const headers = await this.getHeaders();
+    if (headers) {
+      let offset = 0;
+      let rows = await this.read({ limit: GOOGLE_SHEETS_ROWS_LIMIT, offset });
+      while (rows.length > 0) {
+        allRows.push(...rows);
+        offset += GOOGLE_SHEETS_ROWS_LIMIT;
+        rows = await this.read({ limit: GOOGLE_SHEETS_ROWS_LIMIT, offset });
+      }
+    }
+    return allRows;
+  }
+
+  async getSheetRows() {
+    const cacheDurationMs = 1000 * 60 * 3; // 3 minutes
+    const contentId = `${this.docId}-${this.sheetId}`;
+    const cacheKey = ["sheetContent", contentId];
+    return objectCache(
+      { objectId: contentId, cacheKey, cacheDurationMs },
+      async () => {
+        return await this.getAllRows();
+      }
+    );
+  }
```

Comment:
I see what you mean about using memory to hold the whole sheet!

I'm also worried about this filling up all of redis and causing problems system-wide (eg: the background tasks stop working).  I don't have a great idea for a better solution other than emulating what we do on the source side: downloading the whole sheet locally as a CSV and then scanning the file until we find the row we want.

Pros:
* doesn't use Redis' memory

Cons:
* probably slower
* increases the drift between local and remove versions 

Maybe using Redis is OK, but I'd want some sort of upper limit on how big the cached object is... perhaps something like `if (rows * columns > 1M) throw new Error('too big') `

Reply:
Wait... this /doesn't/ store everything in redis! that's just the method to reload the table once the key expires!  @pauloouriques were you instead worried about  `this.getAllRows()` in general possibly filling up Node's memory (rather than Redis')?

Reply:
It still has the `getAllRows` but now with no cache.